// uxcore-menu/build/Menu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _rcMenu = require('rc-menu');

var _rcMenu2 = _interopRequireDefault(_rcMenu);

var _cssAnimation = require('css-animation');

var _cssAnimation2 = _interopRequireDefault(_cssAnimation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Menu Component for uxcore
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @author vincent.bian
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                *
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Copyright 2014-2015, Uxcore Team, Alinw.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * All rights reserved.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


/* eslint-disable no-param-reassign */

var Menu = function (_React$Component) {
  _inherits(Menu, _React$Component);

  function Menu() {
    _classCallCheck(this, Menu);

    return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
  }

  Menu.prototype.animate = function animate(node, show, done) {
    var prefixCls = this.props.prefixCls;

    var height = void 0;
    return (0, _cssAnimation2["default"])(node, prefixCls + '-collapse', {
      start: function start() {
        if (!show) {
          node.style.height = node.offsetHeight + 'px';
          node.style.opacity = 0;
        } else {
          height = node.offsetHeight;
          node.style.height = 0;
          node.style.opacity = 1;
          // node.style.transform = 'translate(-10px, 0)';
        }
      },
      active: function active() {
        node.style.height = (show ? height : 0) + 'px';
        // node.style.transform = `translate(${show ? '0' : '-10px'}, 0)`;
      },
      end: function end() {
        node.style.height = '';
        done();
      }
    });
  };

  Menu.prototype.render = function render() {
    var _this2 = this;

    var me = this;
    var openAnimation = {
      enter: function enter(node, done) {
        return me.animate(node, true, done);
      },
      appear: function appear(node, done) {
        return me.animate(node, true, done);
      },
      leave: function leave(node, done) {
        return me.animate(node, false, done);
      }
    };
    var onOpenChange = function onOpenChange(openKeys) {
      if (_this2.props.onOpenChange) {
        _this2.props.onOpenChange(openKeys);
      }

      if (_this2.props.onOpen) {
        _this2.props.onOpen({ openKeys: openKeys });
      }

      if (_this2.props.onClose) {
        _this2.props.onClose({ openKeys: openKeys });
      }
    };

    if (this.props.mode === 'inline') {
      return _react2["default"].createElement(_rcMenu2["default"], _extends({}, this.props, { openAnimation: openAnimation, onOpenChange: onOpenChange }));
    }
    return _react2["default"].createElement(_rcMenu2["default"], _extends({}, this.props, { onOpenChange: onOpenChange }));
  };

  return Menu;
}(_react2["default"].Component);

Menu.defaultProps = {
  prefixCls: 'kuma-menu',
  openAnimation: '',
  inlineIndent: 14
};
Menu.propTypes = {
  mode: _propTypes2["default"].string,
  prefixCls: _propTypes2["default"].string,
  onOpen: _propTypes2["default"].func,
  onClose: _propTypes2["default"].func
};

Menu.SubMenu = _rcMenu.SubMenu;
Menu.Item = _rcMenu.Item;
Menu.Divider = _rcMenu.Divider;

exports["default"] = Menu;
module.exports = exports['default'];
