// uxcore-validator/build/Validator.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

/**
 * Created by xy on 15/4/16.
 */

var Validator = {};

var PATTERN = {
    EMAIL: /^([a-z0-9_\.-]+)@([\da-z\.-]+)\.([a-z\.]{2,6})$/,
    URL: /^(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?$/,
    HEX: /^#?([a-f0-9]{6}|[a-f0-9]{3})$/i,
    NUM: /^((-?\d+\.\d+)|(-?\d+)|(-?\.\d+))$/,
    IDCARD: /(^\d{15}$)|(^\d{17}([0-9]|X)$)/i,
    CNMOBILE: /^(0|86|17951)?(13[0-9]|15[012356789]|17[0678]|18[0-9]|14[57])[0-9]{8}$/
};

Validator.isNotEmpty = function (value) {
    // empty means empty string, empty array, empty object & null & undefined
    if (typeof value == "string") {
        return value.length !== 0;
    } else if ((typeof value === 'undefined' ? 'undefined' : _typeof(value)) == 'object') {
        if (value instanceof Array) {
            return value.length !== 0;
        } else {
            var i = 0;
            for (var key in value) {
                i++;
            }
            return !!i;
        }
    } else if (typeof value == 'number') {
        return true;
    } else {
        return !!value;
    }
};

Validator.isNum = function (value) {
    return PATTERN.NUM.test(value);
};

Validator.isInt = function (value) {
    return Validator.isNum(value) && parseInt(value) == value;
};

Validator.isDecimal = function (value) {
    return Validator.isNum(value) && !Validator.isInt(value);
};

Validator.isArray = function (value) {
    return Array.isArray(value);
};

Validator.isRegExp = function (value) {
    if (value instanceof RegExp) {
        return true;
    }
    try {
        return !!new RegExp(value);
    } catch (e) {
        return false;
    }
};

Validator.isObject = function (value) {
    return (typeof value === 'undefined' ? 'undefined' : _typeof(value)) === 'object' && !Validator.isArray(value);
};

Validator.isFunc = function (value) {
    return typeof value === 'function';
};

Validator.isEmail = function (value) {
    return typeof value === 'string' && PATTERN.EMAIL.test(value);
};

Validator.isUrl = function (value) {
    return typeof value === 'string' && PATTERN.URL.test(value);
};

Validator.isHex = function (value) {
    return typeof value === 'string' && PATTERN.HEX.test(value);
};

Validator.isIdCard = function (value) {
    return typeof value === 'string' && PATTERN.IDCARD.test(value);
};

Validator.isCNMobile = function (value) {
    return typeof value === 'string' && PATTERN.CNMOBILE.test(value);
};
module.exports = Validator;
