// uxcore-crumb-item/build/prototype.js
'use strict';

var React = require('react');

var _require = require('engine');

var Bundle = _require.Bundle;

var _require2 = require('engine-utils');

var BoolSetter = _require2.BoolSetter;
var TextSetter = _require2.TextSetter;
var ChoiceSetter = _require2.ChoiceSetter;
var JsonSetter = _require2.JsonSetter;
var NumberSetter = _require2.NumberSetter;


module.exports = Bundle.createPrototype({
    title: "面包屑项目",
    category: null,
    icon: require("./logo.svg"), // todo: require("./logo.svg"),
    componentName: "CrumbItem",
    canHovering: false,
    canSelecting: false,
    canDraging: false,
    isInline: true,
    isContainer: false,
    canDropto: true,
    canDroping: false,
    configure: [{
        name: "href",
        title: "链接",
        defaultValue: "",
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }, {
        name: "children",
        title: "文案",
        defaultValue: '链接',
        required: false,
        fieldStyle: "block",
        fieldCollapsed: false,
        setter: React.createElement(TextSetter, { multiline: true, rows: 2 })
    }]
});
