// uxcore-table/build/CollapsedButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _uxcoreButton = require('uxcore-button');

var _uxcoreButton2 = _interopRequireDefault(_uxcoreButton);

var _uxcoreDropdown = require('uxcore-dropdown');

var _uxcoreDropdown2 = _interopRequireDefault(_uxcoreDropdown);

var _uxcoreMenu = require('uxcore-menu');

var _uxcoreMenu2 = _interopRequireDefault(_uxcoreMenu);

var _classnames4 = require('classnames');

var _classnames5 = _interopRequireDefault(_classnames4);

var _i18n = require('./i18n');

var _i18n2 = _interopRequireDefault(_i18n);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

/**
 * only consider the button style and how to collapse
 */

var CollapsedButtonItem = function CollapsedButtonItem() {};

var CollapsedButton = function (_React$Component) {
  _inherits(CollapsedButton, _React$Component);

  function CollapsedButton(props) {
    _classCallCheck(this, CollapsedButton);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.state = {
      dropdownVisible: false
    };
    _this.handleDropdownVisibleChange = _this.handleDropdownVisibleChange.bind(_this);
    _this.handleMoreClick = _this.handleMoreClick.bind(_this);
    _this.refCallback = [];
    return _this;
  }

  CollapsedButton.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    if (_react2["default"].Children.count(nextProps.children) < parseInt(nextProps.maxLength, 10)) {
      this.setState({
        dropdownVisible: false
      });
    }
  };

  CollapsedButton.prototype.componentDidUpdate = function componentDidUpdate() {
    var _props = this.props,
        children = _props.children,
        maxLength = _props.maxLength;

    if (this.state.dropdownVisible && parseInt(maxLength, 10) === 1 && _react2["default"].Children.count(children) > 1) {
      var dropdownDOMNode = this.dropdownInstance.getPopupDomNode();
      /* eslint-disable react/no-find-dom-node */
      var triggerDOMnode = _reactDom2["default"].findDOMNode(this.triggerInstance);
      /* eslint-enable react/no-find-dom-node */
      dropdownDOMNode.style.minWidth = (triggerDOMnode || this.triggerInstance).offsetWidth + 'px';
    }
  };

  CollapsedButton.prototype.saveRef = function saveRef(refName) {
    var me = this;
    if (me.refCallback[refName]) {
      return me.refCallback[refName];
    }
    me.refCallback[refName] = function (c) {
      me[refName] = c;
      return false;
    };
    return me.refCallback[refName];
  };

  CollapsedButton.prototype.handleDropdownVisibleChange = function handleDropdownVisibleChange(visible) {
    var me = this;
    me.setState({
      dropdownVisible: visible
    });
  };

  CollapsedButton.prototype.handleMenuItemClick = function handleMenuItemClick(action, e) {
    if (action.props.disabled) {
      return;
    }
    if (action.props.onClick) {
      action.props.onClick(e);
    }
    this.setState({
      dropdownVisible: false
    });
  };

  CollapsedButton.prototype.handleMoreClick = function handleMoreClick(e) {
    e.preventDefault();
    var me = this;
    me.setState({
      dropdownVisible: !me.state.dropdownVisible
    });
  };

  CollapsedButton.prototype.renderItem = function renderItem(item, index) {
    var me = this;
    var type = me.props.type;

    var itemProps = {
      className: 'action',
      key: index,
      onClick: item.props.disabled ? function () {} : item.props.onClick
    };
    if (type === 'button') {
      return _react2["default"].createElement(
        _uxcoreButton2["default"],
        _extends({}, itemProps, {
          type: item.props.buttonType || (index === 0 ? 'outline' : 'secondary'),
          size: 'small',
          disabled: !!item.props.disabled
        }),
        item.props.children
      );
    }
    return _react2["default"].createElement(
      'a',
      _extends({}, itemProps, {
        className: (0, _classnames5["default"])({
          action: true,
          disabled: !!item.props.disabled,
          'first-action': index === 0
        })
      }),
      item.props.children
    );
  };

  CollapsedButton.prototype.renderMore = function renderMore(actions) {
    var _classnames;

    if (actions.length === 0) {
      return null;
    }
    var me = this;
    var menu = _react2["default"].createElement(
      _uxcoreMenu2["default"],
      null,
      actions.map(function (action, index) {
        return _react2["default"].createElement(
          _uxcoreMenu2["default"].Item,
          { key: index, disabled: !!action.props.disabled },
          _react2["default"].createElement(
            'a',
            { onClick: me.handleMenuItemClick.bind(me, action) },
            action.props.children
          )
        );
      })
    );

    var offsetY = me.props.type === 'button' ? 0 : -20;
    var dropdownOptions = {
      key: 'icon',
      overlay: menu,
      trigger: ['click'],
      align: {
        offset: [0, offsetY]
      },
      visible: me.state.dropdownVisible,
      overlayClassName: (0, _classnames5["default"])((_classnames = {}, _defineProperty(_classnames, me.props.prefixCls + '-collapsed-button-more-dropdown', true), _defineProperty(_classnames, me.props.prefixCls + '-collapsed-button-more-link-dropdown', me.props.type === 'link'), _classnames)),
      onVisibleChange: me.handleDropdownVisibleChange
    };
    var content = _react2["default"].createElement(
      'span',
      null,
      _react2["default"].createElement(
        'span',
        { className: me.props.prefixCls + '-collapsed-button-text' },
        _i18n2["default"][me.props.locale].more
      ),
      _react2["default"].createElement('i', {
        className: (0, _classnames5["default"])(_defineProperty({
          'kuma-icon': true,
          'kuma-icon-triangle-down': !me.state.dropdownVisible,
          'kuma-icon-triangle-up': me.state.dropdownVisible
        }, me.props.prefixCls + '-collapsed-icon-active', me.state.dropdownVisible))
      })
    );
    if (me.props.type === 'button') {
      return _react2["default"].createElement(
        _uxcoreDropdown2["default"],
        dropdownOptions,
        _react2["default"].createElement(
          _uxcoreButton2["default"],
          { className: 'action', type: 'secondary', size: 'small' },
          content
        )
      );
    }
    return _react2["default"].createElement(
      _uxcoreDropdown2["default"],
      dropdownOptions,
      _react2["default"].createElement(
        'a',
        { className: 'action more-action', onClick: me.handleMoreClick },
        content
      )
    );
  };

  CollapsedButton.prototype.renderHoverMenu = function renderHoverMenu() {
    var _classnames3;

    var me = this;
    var _me$props = me.props,
        children = _me$props.children,
        type = _me$props.type;

    var trigger = void 0;
    var options = [];
    _react2["default"].Children.forEach(children, function (child, index) {
      if (index === 0) {
        var triggerContent = _react2["default"].createElement(
          'span',
          null,
          _react2["default"].createElement(
            'span',
            { className: me.props.prefixCls + '-collapsed-button-text' },
            child.props.children
          ),
          _react2["default"].createElement('i', {
            className: (0, _classnames5["default"])({
              'kuma-icon': true,
              'kuma-icon-triangle-down': !me.state.dropdownVisible,
              'kuma-icon-triangle-up': me.state.dropdownVisible
            })
          })
        );
        if (type === 'button') {
          trigger = _react2["default"].createElement(
            _uxcoreButton2["default"],
            {
              className: 'action',
              type: 'outline',
              size: 'small',
              ref: me.saveRef('triggerInstance')
            },
            triggerContent
          );
        } else {
          trigger = _react2["default"].createElement(
            'a',
            { className: 'action', ref: me.saveRef('triggerInstance') },
            triggerContent
          );
        }
      }
      options.push(_react2["default"].createElement(
        _uxcoreMenu2["default"].Item,
        { key: index, disabled: !!child.props.disabled },
        _react2["default"].createElement(
          'a',
          {
            onClick: me.handleMenuItemClick.bind(me, child)
          },
          child.props.children
        )
      ));
    });
    var menu = _react2["default"].createElement(
      _uxcoreMenu2["default"],
      null,
      options
    );

    var offsetY = type === 'button' ? -33 : -45;

    var dropdownOptions = {
      key: 'icon',
      overlay: menu,
      transitionName: '',
      ref: me.saveRef('dropdownInstance'),
      trigger: ['hover'],
      align: {
        offset: [0, offsetY]
      },
      visible: me.state.dropdownVisible,
      overlayClassName: (0, _classnames5["default"])((_classnames3 = {}, _defineProperty(_classnames3, me.props.prefixCls + '-collapsed-button-more-dropdown', true), _defineProperty(_classnames3, me.props.prefixCls + '-collapsed-button-more-link-dropdown', type === 'link'), _classnames3)),
      onVisibleChange: me.handleDropdownVisibleChange
    };

    return _react2["default"].createElement(
      _uxcoreDropdown2["default"],
      dropdownOptions,
      trigger
    );
  };

  CollapsedButton.prototype.render = function render() {
    var me = this;
    var _me$props2 = me.props,
        children = _me$props2.children,
        maxLength = _me$props2.maxLength,
        type = _me$props2.type;

    var buttons = [];
    var options = [];
    if (parseInt(maxLength, 10) === 1 && _react2["default"].Children.count(children) > 1) {
      return _react2["default"].createElement(
        'div',
        null,
        me.renderHoverMenu()
      );
    }
    if (_react2["default"].Children.count(children) <= parseInt(maxLength, 10)) {
      _react2["default"].Children.forEach(children, function (item, index) {
        if (index !== 0 && type === 'link') {
          buttons.push(_react2["default"].createElement(
            'span',
            { key: 'button' + index, className: me.props.prefixCls + '-collapsed-button-split-line' },
            '|'
          ));
        }
        buttons.push(me.renderItem(item, index));
      });
    } else {
      _react2["default"].Children.forEach(children, function (item, index) {
        if (index < parseInt(maxLength, 10) - 1) {
          buttons.push(me.renderItem(item, index));
          if (type === 'link') {
            buttons.push(_react2["default"].createElement(
              'span',
              { key: 'button' + index, className: me.props.prefixCls + '-collapsed-button-split-line' },
              '|'
            ));
          }
        } else {
          options.push(item);
        }
      });
    }

    return _react2["default"].createElement(
      'div',
      null,
      buttons,
      me.renderMore(options)
    );
  };

  return CollapsedButton;
}(_react2["default"].Component);

CollapsedButton.displayName = 'CollapsedButton';

CollapsedButton.Item = CollapsedButtonItem;

CollapsedButton.propTypes = {
  children: _propTypes2["default"].any,
  locale: _propTypes2["default"].string,
  maxLength: _propTypes2["default"].number,
  onClick: _propTypes2["default"].func
};
CollapsedButton.defaultProps = {
  maxLength: 3,
  onClick: function onClick() {},
  type: 'link'
};

exports["default"] = CollapsedButton;
module.exports = exports['default'];
